writer/ynab/ynab.go

```diff
@@ -19,74 +18,115 @@ import (
 const maxMemoSize int = 200  // Max size of memo field in YNAB API
 const maxPayeeSize int = 100 // Max size of payee field in YNAB API
 
-func BulkWriter(cfg ynabber.Config, t []ynabber.Transaction) error {
+var space = regexp.MustCompile(`\s+`) // Matches all whitespace characters
+
+// Ytransaction is a single YNAB transaction
+type Ytransaction struct {
+	AccountID string `json:"account_id"`
+	Date      string `json:"date"`
+	Amount    string `json:"amount"`
+	PayeeName string `json:"payee_name"`
+	Memo      string `json:"memo"`
+	ImportID  string `json:"import_id"`
+}
+
+// Ytransactions is multiple YNAB transactions
+type Ytransactions struct {
+	Transactions []Ytransaction `json:"transactions"`
+}
 
-	budgetID, found := os.LookupEnv("YNAB_BUDGETID")
-	if !found {
-		return fmt.Errorf("env variable YNAB_BUDGETID: %w", ynabber.ErrNotFound)
+// importIDMaker tries to return a unique YNAB import ID to avoid duplicate
+// transactions.
+func importIDMaker(cfg ynabber.Config, y Ytransaction) (string, error) {
+	// Version 1 method for generating YNAB compliant import ID, output example:
+	// YBBR:-741000:2021-02-18:92f2beb1
+	v1 := func(y Ytransaction) string {
+		hash := sha256.Sum256([]byte(y.Memo))
+		return fmt.Sprintf("YBBR:%s:%s:%x", y.Amount, y.Date, hash[:2])
 	}
-	token, found := os.LookupEnv("YNAB_TOKEN")
-	if !found {
-		return fmt.Errorf("env variable YNAB_TOKEN: %w", ynabber.ErrNotFound)
+
+	// Version 2 method for generating YNAB import IDs. This is superior to v1
+	// because it uses the entire transaction to generate a hash.
+	v2 := func(y Ytransaction) (string, error) {
+		transaction, err := json.Marshal(y)
+		if err != nil {
+			return "", fmt.Errorf("failed to marshal YNAB transaction: %s: %w", y, err)
+		}
+		hash := sha256.Sum256(transaction)
+		return fmt.Sprintf("YBBR:%x", hash)[:32], nil
 	}
```

Comment:
@hpernu: we might have a problem with using the entire transaction. If we do that any change to the transaction (memo, payee, etc.) will cause duplicate transactions in YNAB. We need to narrow it down to something that makes for better uniqueness while retaining some abstraction from the specifics of the transaction that might change.